### ISSUE TITLE
Discussion: support the use of distributed filesystems

### ISSUE THREAD

**Issue by jorikvankemenade** [archivematica/Issues#988]
**Please describe the problem you'd like to be solved**
The [scaling Archivematica documentation](https://www.archivematica.org/en/docs/archivematica-1.10/admin-manual/installation-setup/customization/scaling-archivematica/#scaling-out-optimising-across-multiple-machine) provides some pointers on how to use Archivematica in a scale-out scenario. Recently I have been experimenting with this by distributing the Archivematica services to different virtual machines and duplicating the MCP Client. Judging by some issues (#564, #613), responses to my [mailinglist thread](https://groups.google.com/forum/#!searchin/archivematica/scaling|sort:date/archivematica/oxxP5SvCKic/JUuMleEbBQAJ), and this [thread by Thomas Edvardson](https://groups.google.com/forum/#!searchin/archivematica/scaling|sort:date/archivematica/g-0Vz4EJvIY/AAGslsAKCwAJ), other people have attempted this in the past as well.

One particular problem I experience has to do with the `sharedDirectory`. By design, the MCP services use the `sharedDirectory` to maintain a shared state. It is assumed that each MCP service can access this shared state and has a *consistent* view of the filesystem. I have been experimenting with several distributed filesystems to host the `sharedDirectory` and so far I had reasonable results using [CephFS](https://docs.ceph.com/docs/master/cephfs/). From their docs:

> The Ceph File System, or CephFS, is a POSIX-compliant file system built on top of Ceph’s distributed object store, RADOS. CephFS endeavors to provide a state-of-the-art, multi-use, highly available, and performant file store for a variety of applications, including traditional use-cases like shared home directories, HPC scratch space, and distributed workflow shared storage.

So far the performance of CephFS has been promising. The overall performance hit seems to be in the order of 10%. However, one problem I can't seem to fix are the occasional failures caused by file system latency. An example of this is that when you run a `move_transfer` on *client_1* and immediately run a `copy_processing_mcp` on *client_2*.  Occasionally the `copy_processing_mcp` job reaches *client_2* before the filesystem is updated. This causes *client_2* to access a non-existing file and fail. These issues seem to be similar to issues described in #589, #612 and #613.

I believe that with growing data sizes and increased awareness in digital preservation this use-case becomes more and more common and needs to be supported better. Since @ThomasEdvardsen and @helenst have posted some of the related issues they might want to participate in this discussion as well. 

**Describe the solution you'd like to see implemented**
I think that Archivematica is built and tested with a local filesystem in mind. This means that each component is assumed to have a *consistent* view of the filesystem. If we add a distributed file system in the mix this is not the case: latencies and outages influence the view of each node in the cluster. Even when using a [highly consistent filesystem](https://docs.ceph.com/docs/master/cephfs/app-best-practices/) we still had some problems. 

In #613 one of the solutions for this problem was to introduce a delay:

> Testing with an artificial  second delay introduced before the file list is read (wellcometrust/archivematica@f6fc673) stops these errors from happening.

This makes sense, 5 seconds is an enormous amount of time (even in distributed file system territory) and should resolve all latency related issues. I think a variation on this might be a good solution. I propose the following:

1. Add a configuration to tell the MCP services that the filesystem is distributed.
2. Add retries and backoffs to filesystem operations.

More specifically I could Imagine that we add something like a [retry decorator](https://pypi.org/project/retry/) that retries on filesystem errors to functions that have been known to fail on distributed filesystems. 

**Describe alternatives you've considered**
One nontrivial option would be to make the MCP services locality aware. Instead of all the MCP services sharing one `sharedDirectory` the MCP server and all clients would only have the local filesystem. Each MCP Client could do a complete transfer from a-z, not requiring any file movements while processing. This does, however, mean that all the operations for a group of files need to be performed on the same MCP Client. Or there should be a mechanism to move the files between nodes during the processing, but this is an expensive operation so it shouldn't happen too often. Another disadvantage is that the filesystem polling used by the MCP server will break. So overall I think that this is too much of an overhaul to the whole pipeline and scheduling logic to be feasible.

There are probably more alternatives and possible solutions. If you think I missed anything or have anything else to add, please do so!

**Additional context**
My tests are based on running with Archivematica 1.9 and 1.10 on CentOS 7. I have used both [eos](https://eos.web.cern.ch/) and [CephFS](https://docs.ceph.com/docs/master/cephfs/) for hosting the `sharedDirectory`. 

---

**For Artefactual use:**

Before you close this issue, you must check off the following:

- [ ] All pull requests related to this issue are properly linked
- [ ] All pull requests related to this issue have been merged
- [ ] A testing plan for this issue has been implemented and passed (testing plan information should be included in the issue body or comments)
- [ ] Documentation regarding this issue has been written and merged
- [ ] Details about this issue have been added to the release notes

**Comment by jorikvankemenade** [comment]
It might be worthwhile to link #1111 here. Even though I am not sure that the problems are completely related, I notice that at this point almost all of the problems I am experiencing using CephFS are basically a manifestation of #1111.  I could see how #1111 would seem to be a filesystem related problem, while in reality, it is not.

**Comment by jorikvankemenade** [comment]
Ran into a new problem today:

![image](https://user-images.githubusercontent.com/7934339/75484495-30b90b80-59a9-11ea-8a25-e485135756e7.png)

This indicates that the file on the filesystem is still being used by some other process or device. I am currently running stress tests so I might be overcommitting the system.  But this problem, together with #1111, gives me the idea that distributed file systems are not the problem perse. I think it could also be a symptom of jobs having small and subtle race conditions (database, filesystem, Gearman) that show up when you have many workers going through the workflow rapidly.

**Comment by jorikvankemenade** [comment]
I have been performance testing for several days now. I have been running with up to 5 MCP Client VM's, every VM with up to 5 MCP Clients per VM. So I have been running with anywhere between 2 and 20 MCP Clients using the same filesystem. All the errors I got could be related to either #752, #1111, and the problem above (opened #1124). 

An important observation here is that the error frequency rose rapidly with many MCP clients per VM running many small transfers. When running less MCP Clients on less, bigger, transfers the processing was error-free. Here you can think of running either 50 photo albums on 5 nodes with 20 clients, or 24 video files on 5 nodes with 10 clients.

Since CephFS itself seems to behave well with Archivematica, I think we can close this issue. If you have any questions or feel that this is an import issue anyways. Feel free to ask or reopen.

**Comment by sevein** [comment]
> Since CephFS itself seems to behave well with Archivematica, I think we can close this issue. If you have any questions or feel that this is an import issue anyways. Feel free to ask or reopen.

How did you arrive to this conclusion?

**Comment by jorikvankemenade** [comment]
The problem I ran into when opening this issue was actually #1111. Even though I didn't notice this at the time, the "file not found problem" is actually in the opposite direction of what you would expect. The file has been moved, but Archivematica tries to access the old location which is retrieved from the database not the filesystem. 

Next to this, I have been running Archivematica on full load for the last week (5 virtual machines with up to 20 MCP Clients + separate VMs for the other services), without any problems besides the ones I have linked in the issue. During this time I have ingested ~1TB (after normalization) of content. So IMHO the scope of this issue is wrong. I don't run into any problems that seem to specifically related to CephFS, let alone distributed filesystems in general. And the issues I do run into, all have more specific issues that mention the usage of CephFS.
